core/server/worker/src/main/java/alluxio/worker/block/SpaceReserver.java

```diff
@@ -59,25 +61,40 @@ public SpaceReserver(BlockWorker blockWorker) {
     Map<String, Long> capOnTiers = blockWorker.getStoreMeta().getCapacityBytesOnTiers();
     long lastTierReservedBytes = 0;
     for (int ordinal = 0; ordinal < mStorageTierAssoc.size(); ordinal++) {
-      PropertyKey tierReservedSpaceProp =
-          PropertyKeyFormat.WORKER_TIERED_STORE_LEVEL_RESERVED_RATIO_FORMAT.format(ordinal);
       String tierAlias = mStorageTierAssoc.getAlias(ordinal);
-      long reservedSpaceBytes =
-          (long) (capOnTiers.get(tierAlias) * Configuration.getDouble(tierReservedSpaceProp));
-      mBytesToReserveOnTiers.put(tierAlias, reservedSpaceBytes + lastTierReservedBytes);
-      lastTierReservedBytes += reservedSpaceBytes;
+      long capOnTier = capOnTiers.get(tierAlias);
+      // HighWatemark defines when to start the space reserving process
+      PropertyKey tierHighWatermarkProp =
+          PropertyKeyFormat.WORKER_TIERED_STORE_LEVEL_HIGH_WATERMARK_RATIO_FORMAT.format(ordinal);
+      long highWatermarkInBytes =
+          (long) (capOnTier * Configuration.getDouble(tierHighWatermarkProp));
+
+      // LowWatemark defines when to stop the space reserving process if started
+      PropertyKey tierLowWatermarkProp =
+          PropertyKeyFormat.WORKER_TIERED_STORE_LEVEL_LOW_WATERMARK_RATIO_FORMAT.format(ordinal);
+      long reservedBytes =
+          (long) (capOnTier - capOnTier * Configuration.getDouble(tierLowWatermarkProp));
+      mHighWaterMarkInBytesOnTiers.put(tierAlias, highWatermarkInBytes);
+      mReservedBytesOnTiers.put(tierAlias, reservedBytes + lastTierReservedBytes);
+      lastTierReservedBytes += reservedBytes;
     }
   }
 
   private void reserveSpace() {
+    Map<String, Long> usedBytesOnTiers = mBlockWorker.getStoreMeta().getUsedBytesOnTiers();
     for (int ordinal = mStorageTierAssoc.size() - 1; ordinal >= 0; ordinal--) {
       String tierAlias = mStorageTierAssoc.getAlias(ordinal);
-      long bytesReserved = mBytesToReserveOnTiers.get(tierAlias);
-      try {
-        mBlockWorker.freeSpace(Sessions.MIGRATE_DATA_SESSION_ID, bytesReserved, tierAlias);
-      } catch (WorkerOutOfSpaceException | BlockDoesNotExistException | BlockAlreadyExistsException
-              | InvalidWorkerStateException | IOException e) {
-        LOG.warn(e.getMessage());
+      long highWatermarkInBytes = mHighWaterMarkInBytesOnTiers.get(tierAlias);
+      long reservedBytes = mReservedBytesOnTiers.get(tierAlias);
+      if (highWatermarkInBytes > reservedBytes
+          && usedBytesOnTiers.get(tierAlias) >= highWatermarkInBytes) {
+        try {
+          mBlockWorker.freeSpace(Sessions.MIGRATE_DATA_SESSION_ID, reservedBytes, tierAlias);
+        } catch (WorkerOutOfSpaceException | BlockDoesNotExistException
+            | BlockAlreadyExistsException | InvalidWorkerStateException | IOException e) {
+          LOG.warn("SpaceReserver failed to free tier {} to {} bytes used",
+              tierAlias, reservedBytes, e);
```

Comment:
NIT: `e` -> `e.toMessage()` for warnings to be consistent with our logging convention.